modal/_container_entrypoint.py

```diff
@@ -395,11 +449,28 @@ async def handle_user_exception(self) -> AsyncGenerator[None, None]:
 
     @contextlib.asynccontextmanager
     async def handle_input_exception(self, input_id, started_at: float) -> AsyncGenerator[None, None]:
+        global _ignore_cancellation
         try:
             yield
         except KeyboardInterrupt:
             raise
+        except InputCancellation:
+            # just skip creating any output for this input and keep going with the next instead
+            # it should have been marked as cancelled already in the backend at this point so it
+            # won't be retried
+            logger.info(f"The current input ({input_id=}) was cancelled by a user request")
+            await self.complete_call(started_at)
+            return
         except BaseException as exc:
+            if isinstance(exc, asyncio.CancelledError) and (
+                str(exc) == INPUT_CANCELLATION_MESSAGE or _ignore_cancellation
+            ):
+                _ignore_cancellation = False  # reset for next cancellations
+                # for async functions, InputCancellation is represented by CancelledError with the INPUT_CANCELLATION_MESSAGE message
```

Comment:
Just once it would be nice to do things the same way for sync / async functions :)

Reply:
Yeah :( The problem here is that if we raise an exception in the main thread like we do for sync code, it would risk being raised inside of the event loop rather than the user function, basically killing the container...
Maybe we should implement our own event loop (/extensions) at some point to facilitate this sort of thing